### PR TITLE
Ckhardin/imx gstreamer

### PR DESCRIPTION
**Pull request by chardin-cpi**
This is the basic port for the "hacked" up IMX Gstreamer - since the open source one doesn't have the iMX8 support baked in yet.

### CODE REVIEW

File: package/freescale-imx/Config.in [hunk 1]
```diff
@@ -109,6 +109,12 @@ if BR2_PACKAGE_FREESCALE_IMX_HAS_VIV_GPU
 source "package/freescale-imx/imx-gpu-g2d/Config.in"
 source "package/freescale-imx/imx-gpu-viv/Config.in"
 source "package/freescale-imx/kernel-module-imx-gpu-viv/Config.in"
+
+source "package/freescale-imx/imx-gstreamer/Config.in"
+source "package/freescale-imx/imx-gst-plugins-base/Config.in"
+source "package/freescale-imx/imx-gst-plugins-bad/Config.in"
+source "package/freescale-imx/imx-gst-plugins-good/Config.in"
+source "package/freescale-imx/imx-gst1.0-plugin/Config.in"
```
Comment (vgahlan-cpi) [inline package/freescale-imx/Config.in:117]: You could possibly skip imx-gst1.0-plugin as this seems obsolete.

**Comment by vgahlan-cpi** [comment]
Somehow it show PR approved. I just add comments.

File: package/freescale-imx/Config.in (hunk 1 above)
Comment (chardin-cpi) [reply package/freescale-imx/Config.in:117]: by open source terms - maybe - but, that is the actual fork for the yocto builds on the imx Gstreamer. This is the gift that NXP delivered is a forked tree with patches in the "base", the "bad", and a few other plugins.

The open source driver that is normally used - is currently an imx.6 only driver (else just software on the imx8) - so, given the situation - it was basically just port the yocto setup into the tree for the time being.

**Comment by chardin-cpi** [comment]
> Couple of questions; Which baseline release tag is used ? if not we should just standardize to NXP release tag. So, expect we should just pick repo tars/hash from: rel_imx_5.4.24_2.1.0
> 
> Current UCB kernel is aligned with tag rel_imx_5.4.3_2.0.0 and at some point we will migrating to rel_imx_5.4.24_2.1.0.
> 
> I think these patches should already be merged with released tag repos. So, with the release tagged repos, see if you really need any of these patches.
> 
> Same is applicable to libdrm-imx package.
> 
> Just retire old packages, pull latest from release tag and update hash accordingly.

The CodeAurora repos are NOT the upstream free desktop repos - they are forked and patch - so, this is a minimal set of patches to compile the NXP forks for the Gstreamer and just patch the libdrm from free desktop to provide the vivante api that NXP added to their fork.

This is the vanilla free desktop recipie

```
GSTREAMER1_VERSION = 1.16.2
GSTREAMER1_SOURCE = gstreamer-$(GSTREAMER1_VERSION).tar.xz
GSTREAMER1_SITE = https://gstreamer.freedesktop.org/src/gstreamer
GSTREAMER1_INSTALL_STAGING = YES
GSTREAMER1_LICENSE_FILES = COPYING
GSTREAMER1_LICENSE = LGPL-2.0+, LGPL-2.1+
```

and this is the NXP fork

```
IMX_GSTREAMER_VERSION = beca1b382febc2ba38a21fcb8f5092209773eb7f
IMX_GSTREAMER_SITE = https://source.codeaurora.org/external/imx/gstreamer
IMX_GSTREAMER_SITE_METHOD = git
IMX_GSTREAMER_INSTALL_STAGING = YES
IMX_GSTREAMER_LICENSE_FILES = COPYING
IMX_GSTREAMER_LICENSE = LGPL-2.0+, LGPL-2.1+
```

They are not the same code and you cann't just drop in one vs the other because NXP patched across libs/modules/apis like in any hacking vendor would because why does anyone care about an interface.

**Comment by vgahlan-cpi** [comment]
> They are not the same code and you cann't just drop in one vs the other because NXP patched across libs/modules/apis like in any hacking vendor would because why does anyone care about an interface.

I see. I was under impression that we can just use NXP complete forks from release tags overlayed in correct order without instead of taking freedesktop version and adding NXP  patches. It seems you are suggesting below source tree are just patches not complete working fork ?

If we overlay all these together, woudl you still need freedesktop version.  This should work for both iMX6/8

https://source.codeaurora.org/external/imx/gstreamer/tree/?h=rel_imx_5.4.24_2.1.0
https://source.codeaurora.org/external/imx/gst-plugins-bad/tree/?h=rel_imx_5.4.24_2.1.0
https://source.codeaurora.org/external/imx/gst-plugins-base/tree/?h=rel_imx_5.4.24_2.1.0
https://source.codeaurora.org/external/imx/gst-plugins-good/tag/?h=rel_imx_5.4.24_2.1.0

libdrm-imx
https://source.codeaurora.org/external/imx/libdrm-imx/tree/?h=rel_imx_5.4.24_2.1.0

**Comment by chardin-cpi** [comment]
> > They are not the same code and you cann't just drop in one vs the other because NXP patched across libs/modules/apis like in any hacking vendor would because why does anyone care about an interface.
> 
> I see. I was under impression that we can just use NXP complete forks from release tags overlayed in correct order without instead of taking freedesktop version and adding NXP patches. It seems you are suggesting below source tree are just patches not complete working fork ?
> 
> If we overlay all these together, woudl you still need freedesktop version. This should work for both iMX6/8
> 
> https://source.codeaurora.org/external/imx/gstreamer/tree/?h=rel_imx_5.4.24_2.1.0
> https://source.codeaurora.org/external/imx/gst-plugins-bad/tree/?h=rel_imx_5.4.24_2.1.0
> https://source.codeaurora.org/external/imx/gst-plugins-base/tree/?h=rel_imx_5.4.24_2.1.0
> https://source.codeaurora.org/external/imx/gst-plugins-good/tag/?h=rel_imx_5.4.24_2.1.0
> 
> libdrm-imx
> https://source.codeaurora.org/external/imx/libdrm-imx/tree/?h=rel_imx_5.4.24_2.1.0

The libdrm is already covered by the single patch - it is very unclear why they even forked that other then to be annoying. Of course, the didn't add in the meson build patches which is what the upstream uses.

As for the others, it is possible to switch the refs for the plain Gstreamer over to them - but, NXP doesn't make claims to work on the Raspberry PI for whatever other target - might be there. I isolated the changes to this for now. I am not a fan of NXP forks - they definitely show a lot of half-ass hackery in their commits.

**Comment by vgahlan-cpi** [comment]
> The libdrm is already covered by the single patch - it is very unclear why they even forked that other then to be annoying. Of course, the didn't add in the meson build patches which is what the upstream uses.
> 
> As for the others, it is possible to switch the refs for the plain Gstreamer over to them - but, NXP doesn't make claims to work on the Raspberry PI for whatever other target - might be there. I isolated the changes to this for now. I am not a fan of NXP forks - they definitely show a lot of half-ass hackery in their commits.

Its guess its Ok if you are willing to maintain the mess of patches overtime in incremental releases. As we are already working on kernel, uboot, SCU and quite a few other NXP forks that we upgrade with NXP based releases, so using direct form may be more manageable thing to do. I am not sure, if running or not running to Raspberry PI solving any of problem set. 

Over to you. Feel free to merge in case you are convinced with the approach.

**Pull request merged**